### PR TITLE
Feature/dig 53247 file upload frontend

### PR DESCRIPTION
**Pull request by jafeld**
Big redesign of ffe-file-upload-react including new styling package ffe-file-upload, new design can be seen here (requires access): https://www.figma.com/file/cn3VFs107XnyxeLo4Du4ny/Fil-Opplasting---Master?node-id=0%3A1

Some examples of the new design: 

ffe-file-upload-1

ffe-file-upload-2

These are breaking changes.

### CODE REVIEW

**Comment by ivarni** [comment]
Travis doesn't seem happy, https://travis-ci.org/SpareBank1/designsystem/builds/582730542#L837, but other than that it looks good to me.

**Comment by jafeld** [comment]
> Travis doesn't seem happy, https://travis-ci.org/SpareBank1/designsystem/builds/582730542#L837, but other than that it looks good to me.

Yeah, this is due to the role="presentation" I removed on the 'drag and drop' receiver element based on comment on the other PR. Unsure what is preferred here :)

**Comment by ivarni** [comment]
Then I would suggest the [region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Region_role) role along with aria-label describing it as a droppable area.

**Review by ivarni — DISMISSED** [review]

**Comment by jafeld** [comment]
> Then I would suggest the [region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Region_role) role along with aria-label describing it as a droppable area.

Based on conversation in PM, we decided to add the role="presentation" back and we'll create new case issue to deal with potential screen-reader problems

File: packages/ffe-file-upload/less/file-upload.less [hunk 1]
```diff
@@ -0,0 +1,337 @@
+@ffe-blue-royal-escape: escape('@{ffe-blue-royal}');
+@ffe-blue-royal-fill: @ffe-blue-royal-escape;
+@ffe-blue-azure-escape: escape('@{ffe-blue-azure}');
+@ffe-blue-azure-fill: @ffe-blue-azure-escape;
+
+.ffe-file-upload {
+    margin-top: 45px;
+    display: -ms-grid;
```
Comment (antidecaf) [inline packages/ffe-file-upload/less/file-upload.less]: Vendor prefixes are added at build time, so no need to include them in the Less files (will also cause stylelint errors/warnings).

File: packages/ffe-file-upload/less/file-upload.less [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            bottom: 5px;
+
+            @media (min-width: @breakpoint-md) {
+                width: 270px;
+                height: 210px;
+                right: 0;
+                bottom: 0;
+            }
+
+            &__hover {
```
Comment (kwltrs) [inline packages/ffe-file-upload/less/file-upload.less]: this should be a modifier, i.e. `--hover`

Comment (kwltrs) [inline packages/ffe-file-upload/less/file-upload.less]: this should be a modifier, i.e. `--hover`

File: packages/ffe-file-upload/less/file-upload.less [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        grid-template-columns: 1fr;
+
+        &-info {
+            display: grid;
+            grid-column: 1;
+            grid-row: 1;
+            grid-template-columns: 1fr;
+            overflow: hidden;
+
+            &-background {
```
Comment (kwltrs) [inline packages/ffe-file-upload/less/file-upload.less]: The level of nesting is rather confusing and hard to follow. Could you change the nesting so that element names are not nested? (i.e. `&__file-item-stencil-info-background` instead of `&__file-item-stencil &-info &-background`)

File: packages/ffe-file-upload/less/file-upload.less [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    &__file-item-error {
+        display: grid;
+        grid-gap: 5px;
+        grid-template-columns: 25px 1fr;
+
+        &-icon {
+            &:extend(.ffe-field-error-message);
+
```
Comment (kwltrs) [inline packages/ffe-file-upload/less/file-upload.less:278]: The definition of `.ffe-field-error-message` is provided by `ffe-form`, but `ffe-form` is not listed as dependency in `package.json`.

File: packages/ffe-file-upload/less/file-upload.less [hunk 5]
```diff
@@ -0,0 +1,307 @@
+@ffe-blue-royal-escape: escape('@{ffe-blue-royal}');
+@ffe-blue-royal-fill: @ffe-blue-royal-escape;
+@ffe-blue-azure-escape: escape('@{ffe-blue-azure}');
+@ffe-blue-azure-fill: @ffe-blue-azure-escape;
+
+.ffe-file-upload {
+    margin-top: 45px;
+    display: grid;
+    max-width: 400px;
+    grid-template-columns: 1fr;
```
Comment (kwltrs) [inline packages/ffe-file-upload/less/file-upload.less:10]: Could you add a line to `README` regarding browser support as this package is using css grids?

File: packages/ffe-file-upload/less/file-upload.less (hunk 2 above)
Comment (kwltrs) [reply packages/ffe-file-upload/less/file-upload.less]: And out of curiosity, does the `:hover` pseudo selector not work here? Something to do with drag and drop?

**Review by kwltrs — CHANGES_REQUESTED** [review]
Looks like a very useful component. Thanks!

File: packages/ffe-file-upload/less/file-upload.less (hunk 2 above)
Comment (jafeld) [reply packages/ffe-file-upload/less/file-upload.less]: > this should be a modifier, i.e. `--hover`

It is a modifier, eg: 'block__component__modifier'

This modifier is only triggered when dragging files over the component to trigger the modified style, and as such I had to make a custom modifier. Other components, eg the 'delete' button, triggers on mouse-over ( the :hover selector ).

File: packages/ffe-file-upload/less/file-upload.less (hunk 3 above)
Comment (jafeld) [reply packages/ffe-file-upload/less/file-upload.less]: > The level of nesting is rather confusing and hard to follow. Could you change the nesting so that element names are not nested? (i.e. `&__file-item-stencil-info-background` instead of `&__file-item-stencil &-info &-background`)

Sure :) I'll give it another look!

File: packages/ffe-file-upload/less/file-upload.less (hunk 4 above)
Comment (jafeld) [reply packages/ffe-file-upload/less/file-upload.less:278]: Nice catch!

File: packages/ffe-file-upload/less/file-upload.less (hunk 1 above)
Comment (jafeld) [reply packages/ffe-file-upload/less/file-upload.less]: Should be fixed now (otherwise Travis will complain anyway :smile:  )

File: packages/ffe-file-upload/less/file-upload.less (hunk 2 above)
Comment (henrikhermansen) [reply packages/ffe-file-upload/less/file-upload.less]: > It is a modifier, eg: 'block__component__modifier'

It should be `block__element--modifier` as described by [BEM](http://getbem.com/naming/).
Thanks for clearing up why `:hover` isn't suited here.

File: packages/ffe-file-upload/package.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    "lint": "stylelint less/*.less",
+    "test": "npm run lint"
+  },
+  "devDependencies": {
+    "@sb1/ffe-core": "^14.0.4",
+    "stylelint": "^9.4.0"
+  },
+  "peerDependencies": {
+    "@sb1/ffe-core": "^14.0.2",
+    "@sb1/ffe-form": "^10.1.1"
```
Comment (henrikhermansen) [inline packages/ffe-file-upload/package.json]: I would assume this would work with `@sb1/ffe-core@15.0.0` and `@sb1/ffe-form@11.0.0` as well?
If so it should be reflected in the peerDependencies.

File: packages/ffe-file-upload-react/src/FileItem.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
+import React from 'react';
+import { number, func, shape, object, string as stringType } from 'prop-types';
+
+const FileItem = props => (
```
Comment (henrikhermansen) [inline packages/ffe-file-upload-react/src/FileItem.js]: Could you please deconstruct props here?
It's nice to be able to see available props right away, and you'll save yourself alot of `props.` throughout the code.

File: packages/ffe-file-upload-react/src/FileItem.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    <li key={props.index}>
```
Comment (henrikhermansen) [inline packages/ffe-file-upload-react/src/FileItem.js]: I don't see a purpose of providing a key here.
If we are mapping an array to several `FileItem`s somewhere else we would rather give a key to each `FileItem`, as that's where React actually needs it.

File: packages/ffe-file-upload/less/file-upload.less (hunk 2 above)
Comment (jafeld) [reply packages/ffe-file-upload/less/file-upload.less]: > > It is a modifier, eg: 'block__component__modifier'
> 
> It should be `block__element--modifier` as described by [BEM](http://getbem.com/naming/).
> Thanks for clearing up why `:hover` isn't suited here.

Ah, right! My bad :)

File: packages/ffe-file-upload-react/src/FileItem.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        {// File is loading
+        !props.file.document && !props.file.error && (
+            <div className="ffe-file-upload__file-item-stencil">
```
Comment (henrikhermansen) [inline packages/ffe-file-upload-react/src/FileItem.js:8]: There are alot of `div`s throughout this component.
I'm no expert on semantics, but could any of these `div`s possibly be something else to add semantics?

@ivarni

File: packages/ffe-file-upload-react/src/FileUpload.js [hunk 10]
```diff
@@ -29,97 +33,103 @@ class FileUpload extends React.Component {
...
-                    className="ffe-file-upload__button"
-                    aria-invalid={String(!!errorMessage)}
-                    onClick={this.triggerUploadFileNativeHandler}
+                <div className="ffe-file-upload__title">{title}</div>
+                {Object.keys(files).length > 0 ? (
+                    <div className="ffe-file-upload__file-items-section">
+                        {Object.keys(files).map((file, index) => (
+                            <FileItem
+                                key={index}
+                                index={index}
```
Comment (henrikhermansen) [inline packages/ffe-file-upload-react/src/FileUpload.js]: `index` isn't needed here.
`key` is recommended to not use index from `.map()`, but rather some value which actually corresponds to the data you want to use. Perhaps simply using `file` here will do?

Edit: if `file` is the file name, we might get duplicates. You could try using `files[file]` instead.

File: packages/ffe-file-upload-react/src/FileUpload.js [hunk 11]
```diff
@@ -29,97 +33,103 @@ class FileUpload extends React.Component {
...
                 >
-                    <span className="ffe-file-upload__button__label-icon" />
-                    <span id={`${id}-label`}>{label}</span>
-                </button>
+                    <div
+                        className={`ffe-file-upload__upload-section-border${
+                            this.state.hover
+                                ? ' ffe-file-upload__upload-section-border__hover'
+                                : ''
+                        }`}
```
Comment (henrikhermansen) [inline packages/ffe-file-upload-react/src/FileUpload.js]: You could rather use the `classnames` package here.

```
className={ classnames(
  'ffe-file-upload__upload-section-border',
  { 'ffe-file-upload__upload-section-border__hover': this.state.hover }
) }
```

The package is used alot throughout Designsystem.

**Review by henrikhermansen — CHANGES_REQUESTED** [review]

File: packages/ffe-file-upload-react/src/FileItem.js (hunk 9 above)
Comment (ivarni) [reply packages/ffe-file-upload-react/src/FileItem.js:8]: I'm not familiar with the correct semantics for file uploads but we discussed the accesibility part of it and decided we'd make a separate issue for that. Since the upload button is inside the drag-and-drop target container it's difficult to pull that off without just providing a second upload button that is hidden for non-screenreaders outside of it and then just hiding everything else for screenreaders. At that point semantics matter a bit less.

But I didn't really want to mix that into this redesign since this PR is already quite big.

File: packages/ffe-file-upload-react/src/FileUpload.js (hunk 10 above)
Comment (ivarni) [reply packages/ffe-file-upload-react/src/FileUpload.js]: There's no issue with using the index as a key as long as you don't plan on changing the sort order.

Comment (ivarni) [reply packages/ffe-file-upload-react/src/FileUpload.js]: Hm.. well actually in this case that can happen since there is functionality to delete a file. If one is deleted from the front of the list then all the other ones will have their index changed so in this case it actually does matter.

I'll go to my corner.

Comment (jafeld) [reply packages/ffe-file-upload-react/src/FileUpload.js]: > `index` isn't needed here.
> `key` is recommended to not use index from `.map()`, but rather some value which actually corresponds to the data you want to use. Perhaps simply using `file` here will do?
> 
> Edit: if `file` is the file name, we might get duplicates. You could try using `files[file]` instead.

File is an object, so even if we have duplicates, they should not be equal, right?

**Review by ivarni — DISMISSED** [review]

**Comment by ivarni** [comment]
@henrikhermansen @kwltrs your requested changes is what's stopping the merge right now. I'm not sure how to deal with them.

**Review by kwltrs — DISMISSED** [review]

**Review by antidecaf — DISMISSED** [review]

File: packages/ffe-file-upload/package.json [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  "repository": {
+    "type": "git",
+    "url": "ssh://git@github.com:SpareBank1/designsystem.git"
+  },
+  "scripts": {
+    "lint": "stylelint less/*.less",
+    "test": "npm run lint"
+  },
+  "devDependencies": {
+    "@sb1/ffe-core": "^14.0.4",
```
Comment (henrikhermansen) [inline packages/ffe-file-upload/package.json]: This is now mismatching your peer dependency. I would suggest perhaps accepting a wider range of versions in peer dependencies.
Perhaps `>=14.0.0` or `14.x || 15.x`?

File: packages/ffe-file-upload-react/src/FileUpload.js (hunk 10 above)
Comment (henrikhermansen) [reply packages/ffe-file-upload-react/src/FileUpload.js]: > File is an object, so even if we have duplicates, they should not be equal, right?

In the prop types comments here it says `files` is "A map of files, indexed by file name", meaning that `file` is not an object, but `files[file]` is.

However, it occurs to me now that of course the `files` object can't have duplicate keys, but how does this data structure work if the user wants to upload two files with identical file names? :thinking:

File: packages/ffe-file-upload/package.json (hunk 12 above)
Comment (jafeld) [reply packages/ffe-file-upload/package.json]: Regarding the file structure, it's by design the files you upload should not contain duplicates, however, the logic enforcing this currently rest in the implementation, which is not ideal.

In the example, if you upload a file with same name, you overwrite the old one.

Comment (henrikhermansen) [reply packages/ffe-file-upload/package.json]: Okay. That is fine by me, just wanted to make sure we are aware of it, and that we've made a conscious decision that it's fine this way :)

**Review by henrikhermansen — DISMISSED** [review]

**Review by henrikhermansen — APPROVED** [review]
